### PR TITLE
Principal stress visulization output

### PR DESCRIPTION
**Pull request by gassmoeller**
As described in the changelog entry. This adds a useful new visualization postprocessor to compute the principal stress axis and values (useful to determine failure modes or maximum differential stress).

Below are two example outputs for simple shear (top) and pure shear (bottom). Largest principal stress directions are lines in black, smallest principal stress directions lines in white, and differential stress is background color. Results are as expected for these benchmarks, i.e. maximum compressive stress 45 degree rotated to simple shear direction in top figure, and aligned with the pure shear direction of compression / y-axis in bottom figure, respectively.

simple_shear
pure_shear

### CODE REVIEW

File: include/aspect/postprocess/visualization/principal_stress.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      /**
+       * A class derived from DataPostprocessor that outputs the principal
+       * stress values and directions, i.e. the eigenvalues and eigenvectors
```
Comment (tjhei) [inline include/aspect/postprocess/visualization/principal_stress.h:39]:
```suggestion
       * stress values and directions, i.e., the eigenvalues and eigenvectors
```

File: source/postprocess/visualization/principal_stress.cc [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+              = (this->get_material_model().is_compressible()
+                 ?
+                 strain_rate - 1./3. * trace(strain_rate) * unit_symmetric_tensor<dim>()
+                 :
+                 strain_rate);
+
+            const double eta = out.viscosities[q];
+
+            // Compressive stress is positive in geoscience applications
+            SymmetricTensor<2,dim> stress = -2 * eta * compressible_strain_rate + in.pressure[q] * unit_symmetric_tensor<dim>();
```
Comment (tjhei) [inline source/postprocess/visualization/principal_stress.cc:126]:
```suggestion
            SymmetricTensor<2,dim> stress = -2. * eta * compressible_strain_rate + in.pressure[q] * unit_symmetric_tensor<dim>();
```

File: source/postprocess/visualization/principal_stress.cc [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(PrincipalStress,
+                                                  "principal stress",
+                                                  "A visualization output object that outputs the "
+                                                  "principal stress values and directions, i.e. the "
```
Comment (tjhei) [inline source/postprocess/visualization/principal_stress.cc:204]:
```suggestion
                                                  "principal stress values and directions, i.e., the "
```

File: source/postprocess/visualization/principal_stress.cc [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                                                  "eigenvalues and eigenvectors of the stress tensor. "
+                                                  "The postprocessor can either operate on the full "
+                                                  "stress tensor, or only deviatoric stress tensor.")
```
Comment (tjhei) [inline source/postprocess/visualization/principal_stress.cc:207]:
```suggestion
                                                  "stress tensor or only on the deviatoric stress tensor.")
```

**Review by tjhei — APPROVED** [review]
Looks good. I could only find a few typos. :-)

**Comment by gassmoeller** [comment]
Thanks, done.

**Comment by tjhei** [comment]
Looks good to me. I will others have a chance to look at it before merging.

**Comment by dansand** [comment]
I was wondering if this should be expanded to account for visco-elasticity, in which case the stored elastic stresses needs to be accounted for. I think it should be quite a simple change. @naliboff ?

**Comment by gassmoeller** [comment]
Good point, I included the elastic stresses if available. They should probably also be included in the other stress postprocessors (probably a different PR).

**Comment by tjhei** [comment]
> probably a different PR

yes.

**Review by tjhei — APPROVED** [review]

File: tests/viscoelastic_bending_beam_principal_stress.prm [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# This test checks whether the viscoelastic bending beam benchmark
+# run successfully. This test extends viscoelastic_bending_beam by
```
Comment (bobmyhill) [inline tests/viscoelastic_bending_beam_principal_stress.prm:2]:
```suggestion
# runs successfully. This test extends viscoelastic_bending_beam by
```

File: tests/viscoelastic_bending_beam_principal_stress.prm [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+# the output of the principal stresses including the elastic stresses.
```
Comment (bobmyhill) [inline tests/viscoelastic_bending_beam_principal_stress.prm:3]:
```suggestion
# also outputting the principal stresses including the elastic stresses.
```

**Review by bobmyhill — APPROVED** [review]
Great PR! I had to really try hard to find anything that wasn't perfect :D

**Comment by gassmoeller** [comment]
But you still found something :-) ... I should place some more obvious typos in the future ^^.
Fixed. Ready to merge?

**Review by naliboff — APPROVED** [review]
Indeed, really useful PR! Unlike @bobmyhill, I could not find any spelling mistakes :-)

**Comment by naliboff** [comment]
@gassmoeller - It looks like you will need to update the bending beam benchmark test.

**Comment by bobmyhill** [comment]
When the test output 

> But you still found something :-) ... I should place some more obvious typos in the future ^^.
> Fixed. Ready to merge?

When the test output is updated, yes, I think this is ready to merge.

**Comment by anne-glerum** [comment]
Very useful postprocessor, thanks! I was wondering, here you compute stress as
`SymmetricTensor<2,dim> stress = -2. * eta * compressible_strain_rate + in.pressure[q] * unit_symmetric_tensor<dim>()`
while in stress.cc, stress is computed as
`const SymmetricTensor<2,dim> stress = 2*eta*compressible_strain_rate +
                                                                     in.pressure[q] * unit_symmetric_tensor<dim>();`
I'm used to seeing `stress = 2*eta*compressible_strain_rate - P*delta_ij`, with `P=-1/3*sigma_ii` for positive pressure under compression. With a sign conversion for positive stresses under compression as you write in the comment, this would be `-2*eta*compressible_strain_rate + P*delta_ij`. So is the stress.cc postprocessor missing a minus sign? (This comment should probably have been an issue instead, but since we're all here :) )

**Comment by naliboff** [comment]
@anne-glerum - Thanks for bringing this up, which jogged my memory about this discussion on the forum:
https://community.geodynamics.org/t/a-question-of-postprocessing-total-stress-tensor/938.

The conclusion of that discussion was that the pressure should have the opposite sign of stress and thus the following in `stress.cc` needs to be changed:
`const SymmetricTensor<2,dim> stress = 2*eta*compressible_strain_rate + in.pressure[q] * unit_symmetric_tensor<dim>();`

There was just never a PR follow-up. Should we make the two post processors consistent and change `stress.cc` to:
`SymmetricTensor<2,dim> stress = -2. * eta * compressible_strain_rate + in.pressure[q] * unit_symmetric_tensor<dim>()` ?

**Comment by anne-glerum** [comment]
Oh yeah, I forgot about that forum post. I can make a PR for `stress.cc` and for `shear_stress.cc`, which should then also be updated from `const SymmetricTensor<2,dim> shear_stress = 2*eta*compressible_strain_rate;` to `const SymmetricTensor<2,dim> shear_stress = -2*eta*compressible_strain_rate;` to be consistent.

**Comment by gassmoeller** [comment]
I updated the test output. Ready from my side.

@anne-glerum: Thanks for taking a look at the other postprocessors! Can you also check surface_stress.cc? That may have the same issue.

**Pull request merged**